kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -166,14 +235,62 @@ private static DataType convertToPhysicalType(
         return logicalType;
     }
 
-    /** Returns the physical name for a given {@link StructField} */
-    public static String getPhysicalName(StructField field) {
-        if (field.getMetadata().contains(COLUMN_MAPPING_PHYSICAL_NAME_KEY)) {
-            return (String) field
-                .getMetadata()
-                .get(COLUMN_MAPPING_PHYSICAL_NAME_KEY);
-        } else {
-            return field.getName();
+    private static boolean allowMappingModeChange(String oldMode, String newMode) {
+        // only upgrade from none to name mapping is allowed
+        return oldMode.equals(newMode) ||
+                (COLUMN_MAPPING_MODE_NONE.equals(oldMode) &&
+                        COLUMN_MAPPING_MODE_NAME.equals(newMode));
+    }
+
+    /**
+     * For each column/field in a {@link Metadata}'s schema, assign an id using the current
+     * maximum id as the basis and increment from there. Additionally, assign a physical name based
+     * on a random UUID or re-use the old display name if the mapping mode is updated on an
+     * existing table.
+     * @param metadata The new metadata to assign ids and physical names to
+     * @param isNewTable whether this is part of a commit that sets the mapping mode on a new table
+     * @return {@link Metadata} with a new schema where ids and physical names have been assigned
+     */
+    private static Metadata assignColumnIdAndPhysicalName(Metadata metadata, boolean isNewTable) {
+        StructType schema = metadata.getSchema();
+        int maxColumnId = Math.max(
+                Integer.parseInt(metadata.getConfiguration()
+                        .getOrDefault(COLUMN_MAPPING_MAX_COLUMN_ID_KEY, "0")),
+                findMaxColumnId(schema));
+        StructType newSchema = new StructType();
+        for (StructField field : schema.fields()) {
+            StructField updatedField = field;
+
+            if (!hasColumnId(field)) {
+                maxColumnId++;
+                updatedField = field.withNewMetadata(FieldMetadata.builder()
+                        .fromMetadata(field.getMetadata())
+                        .putLong(COLUMN_MAPPING_ID_KEY, maxColumnId)
+                        .build());
+            }
+            if (!hasPhysicalName(field)) {
+                // re-use old display names as physical names when a table is updated
+                String physicalName = isNewTable ? "col-" + UUID.randomUUID() : field.getName();
+                updatedField = updatedField.withNewMetadata(FieldMetadata.builder()
+                        .fromMetadata(updatedField.getMetadata())
+                        .putString(COLUMN_MAPPING_PHYSICAL_NAME_KEY, physicalName)
+                        .build());
+            }
+            newSchema = newSchema.add(updatedField);
         }
+
+        return metadata.withNewSchema(newSchema);
```

Comment:
also don't we need to store the the new max id in the metadata?

Reply:
ah yes, I forgot to push the commit where this is written